server/packetio.go

```diff
@@ -96,26 +98,31 @@ func (p *packetIO) readOnePacket() ([]byte, error) {
 
 	length := int(uint32(header[0]) | uint32(header[1])<<8 | uint32(header[2])<<16)
 
-	data := make([]byte, length)
-	if p.readTimeout > 0 {
-		if err := p.bufReadConn.SetReadDeadline(time.Now().Add(p.readTimeout)); err != nil {
-			return nil, err
-		}
-	}
-	if _, err := io.ReadFull(p.bufReadConn, data); err != nil {
+	// data := make([]byte, length)
+	// if p.readTimeout > 0 {
+	// 	if err := p.SetReadTimeout(p.readTimeout); err != nil {
+	// 		return nil, err
+	// 	}
+	// }
+
+	data, err := r.Next(length)
+
+	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return data, nil
+	buf := make([]byte, length)
+	copy(buf, data)
```

Comment:
hi, `buf := r.ReadBinary(length)` can replace `r.Next && copy` here, which may reduce copy(because of linked node)